### PR TITLE
ARM: issue with constant islands and large switch instructions

### PR DESCRIPTION
**Pull request by xavierleroy**
This is an alternate fix for the issue reported in #994 .

Additionally, it fixes another potential issue with switch instructions whereas the size of the instruction encoding could have been underestimated.

### CODE REVIEW

**Comment by xavierleroy** [comment]
We need to decide quicky whether this goes in 4.05 or is kept for future work.  @mshinwell : an opinion?

**Comment by mshinwell** [comment]
I'm planning to read this one soon, so we should be able to get this into 4.05.

File: asmcomp/arm/emit.mlp [hunk 1]
```diff
@@ -796,30 +801,48 @@ let emit_instr i =
...
+        emit_literals();
+        `{emit_label lbl}:\n`;
+        0
+      end else
+        ninstr in
+    let n = emit_instr i in
+    emit_all (ninstr' + n) (has_fallthrough i.desc) i.next
   end
 
+
```
Comment (mshinwell) [inline asmcomp/arm/emit.mlp:845]: Unrelated whitespace change

File: asmcomp/arm/emit.mlp [hunk 2]
```diff
@@ -862,7 +885,7 @@ let fundecl fundecl =
     end
   end;
   `{emit_label !tailrec_entry_point}:\n`;
-  emit_all 0 fundecl.fun_body;
+  emit_all 0 true fundecl.fun_body;
```
Comment (mshinwell) [inline asmcomp/arm/emit.mlp:888]: I think labelling `~fallthrough` would be clearer.

File: asmcomp/arm/emit.mlp [hunk 3]
```diff
@@ -796,30 +801,48 @@ let emit_instr i =
...
-      emit_all 0 i.next
-    end else if !num_literals != 0 && ninstr' >= limit then begin
-      let lbl = new_label() in
-      `	b	{emit_label lbl}\n`;
-      emit_literals();
-      `{emit_label lbl}:\n`;
-      emit_all 0 i.next
-    end else
-      emit_all ninstr' i.next
+    let limit = limit - !num_literals - max_instruction_size i in
```
Comment (mshinwell) [inline asmcomp/arm/emit.mlp]: The name `num_literals` is confusing since it sounds like, well, the number of literals---but actually it's the number of words they occupy.  Could we rename this?

File: asmcomp/arm/emit.mlp [hunk 4]
```diff
@@ -796,30 +801,48 @@ let emit_instr i =
...
+   in 32-bit words. *)
+
+let max_instruction_size i =
+  match i.desc with
+  | Lswitch jumptbl ->
+      if !arch > ARMv6 && !thumb
+      then 1 + (Array.length jumptbl + 1) / 2 + Array.length jumptbl
+      else 2 + Array.length jumptbl
+  | _ ->
+      4   (* conservative upper bound *)
```
Comment (mshinwell) [inline asmcomp/arm/emit.mlp]: This number doesn't appear to be conservative, e.g. for `Iextcall`.  Could you double-check all the cases?
Maybe there should be an assertion in `emit_all` that `emit_instr` doesn't return a number greater than what `max_instruction_size` returned.

(Aside: problems like this make me increasingly think that the right way to do code emitters is via a DSL, where absolutely everything including layout is described; then it's easier to do things such as measuring.)

**Review by mshinwell — CHANGES_REQUESTED** [review]

**Comment by xavierleroy** [comment]
I fixed `max_instruction_size`, erring on the side of caution, and renamed `num_literals`.  This is good to go, in my opinion.

**Comment by xavierleroy** [comment]
The failing Travis check is unrelated (camlp4 issue).  

Is this good to go?

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
I still reckon there should be an assertion that the max instruction size wasn't an underestimate, but if you don't want to do that, OK.

**Comment by xavierleroy** [comment]
Squashed and merged in trunk, commit 07b28ed .
